packages/financial-templates-lib/src/clients/OptimisticOracleClient.js

```diff
@@ -73,53 +100,116 @@ class OptimisticOracleClient {
+        requester: event.returnValues.requester,
+        identifier: this.hexToUtf8(event.returnValues.identifier),
+        timestamp: event.returnValues.timestamp,
+        currency: event.returnValues.currency,
+        reward: event.returnValues.reward,
+        finalFee: event.returnValues.finalFee
+      };
+    });
+
+    // Store proposals that have NOT been disputed:
+    let undisputedProposals = proposalEvents
+      .filter(event => {
+        const key = this._getPriceRequestKey(event);
+        const hasDispute = disputeEvents.find(disputeEvent => this._getPriceRequestKey(disputeEvent) === key);
+        return hasDispute === undefined;
+      })
+      .map(event => {
+        return {
+          requester: event.returnValues.requester,
+          proposer: event.returnValues.proposer,
+          identifier: this.hexToUtf8(event.returnValues.identifier),
+          timestamp: event.returnValues.timestamp,
+          proposedPrice: event.returnValues.proposedPrice,
+          expirationTimestamp: event.returnValues.expirationTimestamp
+        };
+      });
+
+    // Filter proposals based on their expiration timestamp:
+    const isExpired = proposal => {
+      return Number(proposal.expirationTimestamp) <= Number(currentTime);
+    };
+    this.expiredProposals = undisputedProposals.filter(proposal => isExpired(proposal));
+    this.undisputedProposals = undisputedProposals.filter(proposal => !isExpired(proposal));
+
+    // Store disputes that were resolved and can be settled:
+    let resolvedDisputeEvents = await Promise.all(
+      disputeEvents.map(async disputeEvent => {
+        try {
+          // When someone disputes an OO proposal, the OO requests a price to the DVM with a re-formatted
+          // ancillary data packet that includes the original requester's information:
+          const stampedAncillaryData = await this.oracle.methods
+            .stampAncillaryData(
+              disputeEvent.returnValues.ancillaryData ? disputeEvent.returnValues.ancillaryData : "0x",
+              disputeEvent.returnValues.requester
             )
-            .call(),
-        {
-          concurrency: WEB3_CALLS_BATCH_SIZE
+            .call();
+
+          // getPrice will return null or revert if there is no price available,
+          // in which case we'll ignore this dispute.
+          let resolvedPrice = revertWrapper(
+            await this.voting.methods
+              .getPrice(disputeEvent.returnValues.identifier, disputeEvent.returnValues.timestamp, stampedAncillaryData)
+              .call({
+                from: this.oracle.options.address
+              })
+          );
+          if (resolvedPrice !== null) {
+            return disputeEvent;
+          }
+        } catch (error) {
+          // No resolved price available, do nothing.
         }
-      )
-    ]);
+      })
+      // Remove undefined entries, marking disputes that did not have resolved prices
+    ).filter(event => event !== undefined);
```

Comment:
I just realized you were using bluebird promises which is why this probably works 👍